firebase-firestore/src/main/java/com/google/firebase/firestore/util/AsyncQueue.java

```diff
@@ -176,97 +176,180 @@ private void markDone() {
   }
 
   /**
-   * The single thread that will be used by the executor. This is created early and managed directly
-   * so that it's possible later to make assertions about executing on the correct thread.
+   * A wrapper around a {@link ScheduledThreadPoolExecutor} class that provides:
+   *
+   * <ul>
+   *   <li>1. Synchronized task scheduling. This is different from function 3, which is about task
+   *       execution in a single thread.
+   *   <li>2. Ability to do soft-shutdown: only critical tasks related to shutting Firestore SDK
+   *       down can be executed once the shutdown process initiated.
+   *   <li>3. Single threaded execution service, no concurrent execution among the `Runnable`s
+   *       scheduled in this Executor.
+   * </ul>
    */
-  private final Thread thread;
+  private class SynchronizedShutdownAwareExecutor implements Executor {
```

Comment:
This seems like a lot of extra complexity to support the notion of tasks being able to execute post shutdown. Do we really need this?

What if we just made the default behavior lenient such that tasks submitted via the usual executor interface and everything but `enqueueAndForgetEvenAfterShutdown` just ignored?

Reply:
It looks like a lot but most of the code actually already existed. I moved them from `AsyncQueue` to nested class `SynchronizedShutdownAwareExecutor`.

`SynchronizedShutdownAwareExecutor` implements `Executor` and replace the instance `AsyncQueue` holds. 

`SynchronizedShutdownAwareExecutor` delegate most of its work to `internalExecutor` which plays the role of previous `AsyncQueue.executor`. 

`SynchronizedShutdownAwareExecutor` adds two things (the real changes):

1. Check for shutdown state before accepting a `runnable`.
2. Synchronization of task scheduling/accepting. This is required to prevent race conditions around `isShuttingDown`. Otherwise some task might be scheduled after `shutdown` is called.

This does what you said in your comment above.

(Hope i made it clearer.)

Reply:
I agree that most of the code already existed -- it's not the volume of code I'm objecting to.

The issue I'm trying to express is that there are now multiple executors in play where there used to be one, and the abstraction is incomplete because we're still reaching into the inner executor. We also need to do so carefully to preserve the correct semantics, but at the call site it's not obvious what's happening (see `executor.internalExecutor` used in `enqueueAndForgetEvenAfterShutdown` -- it's not obvious to the reader that this is necessary to bypass the shutdown checks that are otherwise in play).

A similar concern is that the threading primitives in here aren't obvious. AsyncQueue doesn't synchronize on itself, but instead synchronizes on its inner executor. This pattern of nonlocal, non-self-contained synchronization is error prone and hard to evaluate for correctness.

What I'm asking for (and don't entirely understand if it's possible) is can we implement the two concerns you're expressing directly in the AsyncQueue without introducing the additional executor?

Reply:
Modified to hide internalExecutor from outside world as we discussed:
1. Stopped using `AsyncQueue.getExecutorForShutdown` from `GrpcCallProvider`, instead just using `Tasks.await` as you suggested.
2. Pull more implementations to `SynchronizedShutdownAwareExecutor` from `AsyncQueue` such that all `synchronized` operation is contained within SynchronizedShutdownAwareExecutor.